### PR TITLE
ARROW-7965: [Python] Refine higher level dataset API

### PR DESCRIPTION
**Pull request by kszucs**
Provides a more intuitive way to construct nested dataset:

```python
# instead of using confusing factory function
dataset([
     factory("s3://old-taxi-data", format="parquet"),
     factory("local/path/to/new/data", format="csv")
])

# let the user to construct a new dataset directly from dataset objects
dataset([ 
    dataset("s3://old-taxi-data", format="parquet"),
    dataset("local/path/to/new/data", format="csv")
])
```

In the future we might want to introduce a new Dataset class which wraps functionality of both the dataset actory and the materialized dataset enabling optimizations over rediscovery of already materialized datasets.

### CODE REVIEW

File: python/pyarrow/_dataset.pyx
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx]: Can you put this in a private attribute? (or is this not visible on the python side because it's a readonly cdef attribute?)

**Comment by jorisvandenbossche** [comment]
That makes the usage indeed a bit easier. But I am wondering: how expensive is "finishing" the factory? As now you are finishing all sub-datasets, just to reuse the non-finished factory later one.

File: python/pyarrow/_dataset.pyx
Comment (kszucs) [reply python/pyarrow/_dataset.pyx]: Yes, but I can rename it to `_factory`

**Comment by kszucs** [comment]
Yeah, that's a downside until we wrap both the DatasetFactory and Dataset with a single class which does the discovery lazily.

File: python/pyarrow/dataset.py
Comment (kszucs) [inline python/pyarrow/dataset.py]: Perhaps we could rename this as `discover()`

**Comment by nealrichardson** [comment]
I agree with the goal here, but I wonder if the solution should perhaps be in C++? That way we wouldn't have to reimplement this in R too.

**Comment by kszucs** [comment]
> I agree with the goal here, but I wonder if the solution should perhaps be in C++? That way we wouldn't have to reimplement this in R too.

cc @bkietz

File: python/pyarrow/dataset.py
Comment (jorisvandenbossche) [reply python/pyarrow/dataset.py]: Isn't that the name with which we started before we had "factory" .. ? ;)

**Comment by jorisvandenbossche** [comment]
If pushing into C++, that will have the same problem of the double cost of the dataset discovery

**Comment by kszucs** [comment]
@jorisvandenbossche in order to reuse the inspected schemas AFAICS we need "cache" them in the C++ instances and return them from `InspectSchemas()` unless we request an explicit re-discovery.

**Comment by jorisvandenbossche** [comment]
Trying to put in words a bit more structured what I said yesterday in the meeting:

Right now, we need to keep a reference to the dataset factories to be able to reuse those factories, since `UnionDatasetFactory` is written to expect a vector of factories. 

But instead of fixing this on the dataset side by holding this reference, can we change the implementation of `UnionDatasetFactory` to actually accepts a vector of datasets instead of a vector of factories?

Looking at the `UnionDatasetFactory` implementation, it doesn't look like that still needs the schema's of all files/fragments of its sub-datasets. Rather, it only needs the finished schema of its sub-datasets (and then it checks if those are compatible / can be unified). But assuming that those sub-datasets would also created with a factory, the schema returned from `child_factory->Inspect()` will be the same as the schema from the finished dataset from that `child_factory`.

**Comment by jorisvandenbossche** [comment]
OK, I wanted to put this in some pseudo python code to explain it better, which lead me to see that my suggestion is right now indeed not possible ;)

The current logic:
```
def UnionDatasetFactory(factories):
    # inspect
    schemas = []
    for fact in factories:
        schemas.append(fact.inspect())
    schema = UnifySchemas(schemas)

    # finish
    datasets = []
    for fact in factories:
        datasets.append(fact.finish(schema))
    return UnionDataset(datasets, schema)
```

The issue that makes it not possible right now to pass a vector of datasets, is that the sub-datasets need to be finished with a potentially different (unified) schema. 
So my question is then: would it be possible to kind of create a new "view" on an existing dataset with a different (but compatible!) schema? In the end, you could create a new Dataset passing all building blocks of the original dataset (the format, filesystem, forest, file_partitions) but with only specifying a different schema?

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py]: It is more meaningful for filesystems as a VERB.

**Comment by jorisvandenbossche** [comment]
@bkietz @fsaintjacques with your deeper knowledge of the C++ dataset code, could you confirm or not whether my idea above is possible: having a UnionDataset constructor taking a vector of Dataset of objects instead of vector of DatasetFactories ?

**Comment by bkietz** [comment]
@jorisvandenbossche this is already supported in C++, but currently the schemas must be identical rather than merely compatible. Creating a view of a dataset with differing schema would probably be straightforward, could you create a JIRA?

**Comment by jorisvandenbossche** [comment]
Thanks for opening the issue!

**Comment by wesm** [comment]
What's the status of this patch? It's still in the 0.17.0 backlog

**Comment by jorisvandenbossche** [comment]
This is blocked by https://github.com/apache/arrow/pull/6721 (which is blocked by a strange R failure, I think).

I added it to the 0.17 milestone, as it it would be nice to get this in, as I noted in the JIRA:

> This depends on ARROW-8164, but if that gets merged quickly, it would be nice tackle this issue for 0.17 since that would enable us to remove `factory()` from the high-level user API (which wasn't there yet in 0.16, so this would avoid it ever being in a released version)

But it is also certainly not a blocker, since the datasets API is still experimental anyway.

**Comment by nealrichardson** [comment]
#6721 has been merged, so this is unblocked

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py]: Outdated.

File: python/pyarrow/_dataset.pyx
Comment (kszucs) [reply python/pyarrow/_dataset.pyx:60]: Removed the whole thing.

File: python/pyarrow/dataset.py
Comment (kszucs) [inline python/pyarrow/dataset.py:523]: Raise more meaningful errors.

File: python/pyarrow/tests/test_dataset.py
Comment (kszucs) [inline python/pyarrow/tests/test_dataset.py:999]: Test the schema unification.

File: python/pyarrow/dataset.py
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py:432]:
```suggestion
            "When passing a list of Datasets, you cannot pass any additional arguments"
```

File: python/pyarrow/dataset.py
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py:26]: Can you import those inside the functions as it was before (or another import that doesn't import the actual names, like `from pyarrow import fs`) ? 

I did that on purpose to avoid adding those names to the namespace. 
(we can discuss adding those here on purpose, but I would keep that for a separate discussion. But we should avoid adding a subset which depends on implementation details)

File: python/pyarrow/dataset.py
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py:394]:
```suggestion
    return filesystem, paths_or_selector
```

File: python/pyarrow/dataset.py
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py:264]: Another option to infer it from the first path? (re-using part of the code from `_ensure_single_source`)

File: python/pyarrow/dataset.py
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py:237]: An option would also be to punt on this one for now (we need to fix this on the C++ anyway, to have an option to not ignore those files when scanning, there is an open JIRA for it). 

I am not sure what the cost is of doing it here upfront in python, but maybe it is fine for `dataset()`, since you can still use the lower level `FileSystemDataset` constructor to avoid it.

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks a lot for this PR! The diff is a bit hard to interpret, but generally looks good to me.

One thing I am not really sure about is the string URI value to specify the `filesystem`. What's the use for that? If you have a URI, just pass it to the source? This seems a needless complication to me, and not something we support elsewhere? (or do we?)

File: python/pyarrow/types.pxi
Comment (jorisvandenbossche) [reply python/pyarrow/types.pxi:1282]: Alternative could also be a `unify_schemas` *function* (as https://github.com/apache/arrow/pull/6890), instead of a class method (eg concat_table is also a function and not a classmethod)

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py:237]: You mean to check this unconditionally for remote filesystems as well?

Comment (kszucs) [reply python/pyarrow/dataset.py:237]: Because we can pass a flag to validate the file paths set to true by default.

Comment (jorisvandenbossche) [reply python/pyarrow/dataset.py:237]: I meant to just leave this to the C++ factory behaviour (which is right now to ignore them, but we are going to fix that on that level -> https://issues.apache.org/jira/browse/ARROW-7673, https://issues.apache.org/jira/browse/ARROW-8283

File: python/pyarrow/_dataset.pyx
Comment (bkietz) [inline python/pyarrow/_dataset.pyx:1247]:
```suggestion
    cdef init(self, const shared_ptr[CDatasetFactory]& sp):
```

File: python/pyarrow/_dataset.pyx
Comment (bkietz) [inline python/pyarrow/_dataset.pyx:1037]: unused?

File: python/pyarrow/dataset.py
Comment (bkietz) [inline python/pyarrow/dataset.py:72]:
```suggestion
    the expression is bound to a dataset having an explicit schema.
```

File: python/pyarrow/dataset.py
Comment (bkietz) [inline python/pyarrow/dataset.py:441]: This logic is already applied in C++ if no schema is passed to CUnionDataset.Make

File: python/pyarrow/dataset.py
Comment (bkietz) [inline python/pyarrow/dataset.py:454]:
```suggestion
    source : path, list of paths, dataset, list of datasets, factory, or list of factories
```

Not sure what's legal in docstrings like this, but:
```suggestion
    source : SingleOrList[path], SingleOrList[dataset], or SingleOrList[factory]
```

File: python/pyarrow/dataset.py
Comment (bkietz) [inline python/pyarrow/dataset.py:458]: mention the list-of-datasets argument case? In general, this section seems like it might be clearer if presented as a definition list keyed by type(source)

**Review by bkietz — CHANGES_REQUESTED** [review]

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py:444]: Great!

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py:457]: I'm not sure that it currently works with factories. I'm checking it before changing the docstring.

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py:444]: Adding tests and updating.

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py:237]: Sure, my intention was to provide better error messages at least locally until this feature is not available in the C++ implementation.

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py]: @bkietz without that the tests are failing.

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py]: Updated with some additional docstrings and examples (shamelessly copied some parts from Joris' [guide](https://github.com/apache/arrow/pull/6779))

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py:264]: It doesn't support passing URIs as file paths. For example `['s3://file', 'file:///file']` wouldn't work. A dataset requires paths from the same filesystem.

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py]: Added `__all__`. I wonder: do we want to expose the low level classes under `pyarrow.dataset`?

Comment (jorisvandenbossche) [reply python/pyarrow/dataset.py]: Adding `__all__` won't help for this I think.

> I wonder: do we want to expose the low level classes under pyarrow.dataset?

Yes I think so, eg kartothek will want to use them

File: python/pyarrow/dataset.py
Comment (jorisvandenbossche) [reply python/pyarrow/dataset.py:300]: Yes, with two different filesystems it won't work (you will get an error then anyway). But if you pass multiple URIs from the same file system, it *could* work?

File: python/pyarrow/dataset.py
Comment (jorisvandenbossche) [reply python/pyarrow/dataset.py:273]: I don't really have a good intuition on how expensive this is, so I am fine with keeping as is (the behaviour might change again though once we fix this on the c++ side)

File: python/pyarrow/dataset.py
Comment (jorisvandenbossche) [reply python/pyarrow/dataset.py:506]: I don't think it needs to work with a factory? (If you are using a factory, you can just call the finish method yourself to get a Dataset)

> SingleOrList[path], SingleOrList[dataset], or SingleOrList[factory]

I think the "plain english" is easier to understand

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py]: Updated.

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py:259]: May we defer this to a follow-up?

Comment (jorisvandenbossche) [reply python/pyarrow/dataset.py:259]: Yes, that's certainly fine

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py:469]: After taking a try, I think the high level dataset function should not support factory objects. If someone creates factory objects it's unlikely that she/he would pass it to dataset rather than manually calling finish.

**Comment by kszucs** [comment]
@github-actions crossbow submit conda-win-vs2015-py36

**Comment by jorisvandenbossche** [comment]
@kszucs can you also respond to my initial review comment at https://github.com/apache/arrow/pull/6505#pullrequestreview-391336071 ?

**Comment by kszucs** [comment]
Master [is failing](https://dev.azure.com/ursa-labs/crossbow/_build/results?buildId=9571&view=logs&j=4c86bc1b-1091-5192-4404-c74dfaad23e7&t=ec99a26b-0264-5e86-36fb-9cfd0ca0f9f3&l=3892) with windows conda builds, let's see whether the issue persists after this change.

**Comment by kszucs** [comment]
> Thanks a lot for this PR! The diff is a bit hard to interpret, but generally looks good to me.
> 
> One thing I am not really sure about is the string URI value to specify the `filesystem`. What's the use for that? If you have a URI, just pass it to the source? This seems a needless complication to me, and not something we support elsewhere? (or do we?)

```python
files = [
    "part0.parquet",
    "part4.parquet",
    "latest.parquet",
]
ds.dataset(files, filesystem='s3://mybucket/prefix')
```
vs.
```python
files = [
    "mybucket/prefix/part0.parquet",
    "mybucket/prefix/part4.parquet",
    "mybucket/prefix/latest.parquet",
]
ds.dataset(files, filesystem=S3FileSystem())
```

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py:259]: created https://issues.apache.org/jira/browse/ARROW-8400

**Comment by kszucs** [comment]
Seems like the windows paths are not working properly, investigating it.

**Comment by kszucs** [comment]
@github-actions crossbow submit conda-win-vs2015-py36

**Comment by jorisvandenbossche** [comment]
> It helps when the user provides multiple files:

OK, I certainly understand the use case now. 
But, I still find it a bit an unorthodox way to specify a filesystem. Isn't this also what SubTreeFileSystem is meant for?

(I would personally perfer to have that in a separate PR, so the discussion about it doesn't disappear in this big PR, there are already a lot of changes here)

**Comment by kszucs** [comment]
> > It helps when the user provides multiple files:
> 
> OK, I certainly understand the use case now.
> But, I still find it a bit an unorthodox way to specify a filesystem. Isn't this also what SubTreeFileSystem is meant for?

Yes, it's just an alternative to construct the subtreefilesystem from an URI.

**Comment by kszucs** [comment]
@github-actions crossbow submit conda-win-vs2015-py36

**Comment by jorisvandenbossche** [comment]
Failures are due to "ignore_prefix" -> "selector_ignore_prefix" rename on the C++ side.

Since all Ursabot python builds are passing, it also seems they are not running any of the parquet or dataset related tests ..

**Comment by kszucs** [comment]
@github-actions crossbow submit conda-win-vs2015-py36

**Comment by kszucs** [comment]
> Failures are due to "ignore_prefix" -> "selector_ignore_prefix" rename on the C++ side.

Updated.

> 
> Since all Ursabot python builds are passing, it also seems they are not running any of the parquet or dataset related tests ..

Yes, but we're not actively maintaining the ursabot builds now.

File: python/pyarrow/dataset.py
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py:520]: Personally, I would keep this as `ignore_prefixes` in the public API. "selector" is a strange thing for users, as they don't necessarily pass that (rather a string path to a directory). I think a note in the explanation that this is only applicable for directories (as you did) is good enough.

File: python/pyarrow/tests/test_dataset.py
Comment (pitrou) [inline python/pyarrow/tests/test_dataset.py:948]: No. This should be:
* `file:///{}` on Windows
* `file://{}` on non-WIndows

You also probably want to make yourself a helper function...

File: python/pyarrow/dataset.py
Comment (pitrou) [inline python/pyarrow/dataset.py:60]: "Reference"

File: python/pyarrow/dataset.py
Comment (pitrou) [inline python/pyarrow/dataset.py:320]: Missing `)` at the end of the error message.

File: python/pyarrow/dataset.py
Comment (pitrou) [inline python/pyarrow/dataset.py:311]: https://docs.python.org/3/library/exceptions.html#IsADirectoryError

File: python/pyarrow/dataset.py
Comment (pitrou) [inline python/pyarrow/dataset.py:317]: `IOError` would be preferrable here.

File: python/pyarrow/dataset.py
Comment (pitrou) [inline python/pyarrow/dataset.py:461]: "Datasets provide"

File: python/pyarrow/dataset.py
Comment (pitrou) [inline python/pyarrow/dataset.py:467]: "predicate"

File: python/pyarrow/dataset.py
Comment (pitrou) [inline python/pyarrow/dataset.py:477]: "or URI"?

**Review by pitrou — COMMENTED** [review]
I only skimmed through this, a few comments.

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py:522]: Agreed.

File: python/pyarrow/dataset.py
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py:222]: Can be for a follow-up JIRA, but we might want to move this helper to `pyarrow.fs`, and use some this also in other places where we accept filesystems? (although I actually don't know if there are already many such places ..)

File: python/pyarrow/dataset.py
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py:284]:
```suggestion
    IsADirectoryError / IOError
```

File: python/pyarrow/tests/test_dataset.py
Comment (jorisvandenbossche) [inline python/pyarrow/tests/test_dataset.py:1198]: What is "pina"? 

(also, part of this test seems to duplicate the test above?)

**Review by jorisvandenbossche — APPROVED** [review]

File: python/pyarrow/dataset.py
Comment (pitrou) [reply python/pyarrow/dataset.py:284]: `IOError` (or `OSError`) alone is sufficient, since it's a superclass of the former.

**Comment by jorisvandenbossche** [comment]
@kszucs can you also add this patch to this branch:

```patch
--- a/python/pyarrow/tests/test_parquet.py
+++ b/python/pyarrow/tests/test_parquet.py
@@ -2490,13 +2490,7 @@ def _assert_dataset_paths(dataset, paths, use_legacy_dataset):
         assert set(map(str, paths)) == {x.path for x in dataset.pieces}
     else:
         paths = [str(path.as_posix()) for path in paths]
-        if hasattr(dataset._dataset, 'files'):
-            assert set(paths) == set(dataset._dataset.files)
-        else:
-            # UnionDataset
-            # TODO(temp hack) remove this branch once ARROW-7965 is in (which
-            # will change this to a FileSystemDataset)
-            assert dataset.read().num_rows == 50
+        assert set(paths) == set(dataset._dataset.files)
```

(that's a clean-up for a hack I added yesterday becuase of a list of paths giving a UnionDataset, which this PR is fixing)

File: python/pyarrow/tests/test_dataset.py
Comment (kszucs) [reply python/pyarrow/tests/test_dataset.py:1198]: Sorry, that a quick placeholder I use. Updating.

File: python/pyarrow/dataset.py
Comment (kszucs) [reply python/pyarrow/dataset.py:478]: added to the dataset() function

File: python/pyarrow/dataset.py
Comment (pitrou) [inline python/pyarrow/dataset.py:282]: You don't need to detail all this. This is pointless cognitive overhead for the user, IMHO. Just document `IOError`.

File: python/pyarrow/tests/test_dataset.py
Comment (pitrou) [inline python/pyarrow/tests/test_dataset.py:84]: Rather `if os.name == 'nt'`. This function should ideally work fine on regular string.

File: python/pyarrow/tests/test_dataset.py
Comment (pitrou) [inline python/pyarrow/tests/test_dataset.py:1228]: How much time does this test take, out of curiosity?

File: python/pyarrow/tests/test_schema.py
Comment (pitrou) [inline python/pyarrow/tests/test_schema.py:659]: Nice!

File: python/pyarrow/types.pxi
Comment (pitrou) [inline python/pyarrow/types.pxi:1620]: "Unify"

File: python/pyarrow/types.pxi
Comment (pitrou) [inline python/pyarrow/types.pxi:1626]: I suppose it would be better if metadata was merged too, but that's probably on the C++ side....

File: python/pyarrow/types.pxi
Comment (pitrou) [inline python/pyarrow/types.pxi:1623]: Perhaps add that two fields with different types will fail merging?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thank you. A few more comments.

File: python/pyarrow/tests/test_dataset.py
Comment (kszucs) [reply python/pyarrow/tests/test_dataset.py:1228]: The whole case takes 1.5s where ~1.43s is consumed by the bucket creation.

Comment (kszucs) [reply python/pyarrow/tests/test_dataset.py:1229]: Either way, removed redundant reads.

**Review by pitrou — APPROVED** [review]

**Comment by kszucs** [comment]
Thanks Joris, Ben, Antoine! Merging.

**Pull request closed without merging**
